src/concurrent-queue.js

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2020 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import Async from 'async';
+import { v4 as uuid } from 'uuid';
+
+import UploadOptionsBase from './upload-options-base';
+
+/**
+ * Creates a new "batch" that has a given number of items in it. When all
+ * items in the batch have been processed, the given callback will be invoked.
+ * @param {number} totalCount Total number of items in the batch.
+ * @param {function} callback Invoked without arguments when all items in the
+ *  batch have processed.
+ * @returns {string} The ID of the new batch.
+ */
+function createBatch(totalCount, callback) {
+    const batchId = uuid();
+    this.batchCounts[batchId] = {
+        totalCount,
+        totalProcessed: 0,
+        callback,
+    };
+    return batchId;
+}
+
+/**
+ * Increments the total number of items that have been processed in a batch.
+ * If all items in the batch are processed, the method will also invoke the
+ * batch's callback.
+ * @param {string} batchId Batch whose processed count will be incremented.
+ */
+function updateBatch(batchId) {
+    this.batchCounts[batchId].totalProcessed += 1;
+
+    const {
+        totalCount,
+        totalProcessed,
+        callback,
+    } = this.batchCounts[batchId];
+    if (totalCount === totalProcessed) {
+        callback();
+    }
+}
+
+/**
+ * A queue that will ensure that a maximum number of items within it will
+ * be processed concurrently. If at the max, additional items added to the
+ * queue will be "blocked" until it's their turn for processing.
+ *
+ * The queue will automatically begin processing when an item is added, and
+ * will restart if previously added items have all finished processing by the
+ * time a new item is added.
+ *
+ * The queue provides the following events:
+ * * error: sent if there is an error while processing one of the items in the
+ *   queue. Event data will include an "item" property containing the item that
+ *   failed (as it was provided when added to the queue), and an "error" property
+ *   containing details of the error.
+ * * emptied: sent when all the items in the queue have been processed. Could
+ *   potentially be called multiple times, depending on the state of the queue
+ *   when additional items are added.
+ */
+export default class ConcurrentQueue extends UploadOptionsBase {
```

Comment:
I'm not sure `UploadOptionsBase` should be a base class share by every other class in this PR.
Is it really an inheritance (aka `ConcurrentQueue` IS A `UploadOptionsBase `)? 
It feels more like `UploadOptionsBase` should be an attribute of the class (rather a composition than an inheritance).
Wdyt?

Reply:
In any case, I think `ConcurrentQueue` should not be in the same inheritance hierarchy as the other uploaders.

Reply:
Agreed that the base class seems more like a mix-in, but I'm unsure of a good way of doing that in javascript. The primary reason for the common base class is so that all classes have access to logging methods (i.e. `logDebug`, `logInfo`, etc.), and certain classes have access to other methods that they need, depending on which dependencies they use.

In the case of `ConcurrentQueue`, it's an `UploadOptionsBase` because it uses upload options to determine how many items it should concurrently process. So when looking at it that way, yes IMO it IS A `UploadOptionsBase`.

That said, you make some good points. I do agree it could be improved and will think on it more. However, it's not a problem I want to try and solve in the context of this changeset.

Reply:
ACK, thanks for clarifying!